projects/baby_driver/test/test_spi_exchange.c

```diff
@@ -0,0 +1,354 @@
+    s_received_data[s_num_messages] = data[i];
+    if (data[i] != 0) s_num_messages++;
+  }
+  *tx_result = false;
+
+  return STATUS_CODE_OK;
+}
+
+static void send_meta_data(uint8_t port, uint8_t mode, uint8_t tx_len, uint8_t rx_len,
+                           uint8_t cs_port, uint8_t cs_pin, uint8_t use_cs, uint32_t baudrate) {
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_SPI_EXCHANGE_METADATA_1,  // id
+                          port, mode,                                  // port, mode
+                          tx_len, rx_len,                              // tx_len, rx_len
+                          cs_port, cs_pin, use_cs);                    // cs_port, cs_pin, use_cs
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  uint8_t baudrate_1 = (baudrate & 0x000000ff);
+  uint8_t baudrate_2 = (baudrate & 0x0000ff00) >> 8;
+  uint8_t baudrate_3 = (baudrate & 0x00ff0000) >> 16;
+  uint8_t baudrate_4 = (baudrate & 0xff000000) >> 24;
+
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_SPI_EXCHANGE_METADATA_2,      // id
+                          baudrate_1, baudrate_2, baudrate_3, baudrate_4,  // baudrate
+                          0, 0, 0);                                        // 3 * uint8 unused
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+}
+
+static void send_first_meta_data(uint8_t port, uint8_t mode, uint8_t tx_len, uint8_t rx_len,
+                                 uint8_t cs_port, uint8_t cs_pin, uint8_t use_cs) {
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_SPI_EXCHANGE_METADATA_1,  // id
+                          port, mode,                                  // port, mode
+                          tx_len, rx_len,                              // tx_len, rx_len
+                          cs_port, cs_pin, use_cs);                    // cs_port, cs_pin, use_cs
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+}
+
+static void clear_received_data_array() {
+  for (uint8_t i = 0; i < s_num_messages; i++) {
+    s_received_data[i] = 0;
+  }
+}
+
+static void test_equal_multiple_messages() {
+  uint8_t mock_array[296];
+  for (int i = 0; i < s_num_messages; i++) {
+    if (i % 8 == 0) {
+      mock_array[i] = BABYDRIVER_MESSAGE_SPI_EXCHANGE_RX_DATA;
+    } else {
+      mock_array[i] = 1;
+    }
+  }
+
+  for (int i = 0; i < s_num_messages; i++) {
+    TEST_ASSERT_EQUAL(mock_array[i], s_received_data[i]);
+  }
+}
+
+static void send_data(uint8_t bytes) {
+  if (bytes == 1)
+    CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_SPI_EXCHANGE_TX_DATA,  // id
+                            1, 0, 0, 0, 0, 0, 0);
+  else if (bytes == 3)
+    CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_SPI_EXCHANGE_TX_DATA,  // id
+                            1, 1, 1, 0, 0, 0, 0);
+  else if (bytes == 7)
+    CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_SPI_EXCHANGE_TX_DATA,  // id
+                            1, 1, 1, 1, 1, 1, 1);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+}
```

Comment:
It seems a bit sketchy for this function to only work for three specific values of `bytes`; it would definitely be better for it to work for 0-7. So that you don't have to use a big huge if statement, try using the plain old `can_transmit` function from can.h: construct a `CanMessage` (from can_msg.h), and use `NULL` for the ack request. (Take a look at the implementation of the `CAN_TRANSMIT_MACRO` for hints.)

Also, please use {curly brackets} for if statements that aren't super short on one line.